asllib/doc/ASLTypingReference.tex

```diff
@@ -2895,260 +3395,397 @@ \subsection{Example}
+  \item $\newty$ is the bit-vector type with width given by the expression
+    $\ewidthp$ and the bitfields given by $\bitfieldsp$, that is, $\TBits(\ewidthp, \bitfieldsp)$
 \end{itemize}
 
 \subsection{Example}
 \VerbatimInput{../tests/ASLTypingReference.t/TypingRule.TBits.asl}
 
-
-
 In the following example, all the uses of bitvector types are valid:
 \CodeSubsection{\TBitsBegin}{\TBitsEnd}{../Typing.ml}
 
+\begin{emptyformal}
+\subsection{Formally}
+\begin{mathpar}
+\inferrule{
+  \tododefine{annotate\_static\_constrained\_integer}(\tenv, \ewidth) \typearrow \ewidthp \OrTypeError\\
+  \annotatebitfield(\tenv, \ewidthp) \typearrow \bitfieldsp \OrTypeError
+}{
+  \annotatetype{\Ignore, \tenv, \TBits(\ewidth, \bitfields)} \typearrow \TBits(\ewidthp, \bitfieldsp)
+}
+\end{mathpar}
+\end{emptyformal}
 
 \isempty{\subsection{Comments}}
 
-
 \section{TypingRule.TTuple \label{sec:TypingRule.TTuple}}
 
 \subsection{Prose}
-Annotating a type~$\tty$ in an environment~\texttt{env} results in a
-rewritten type~\texttt{new\_ty} and all of the following apply:
+All of the following apply:
 \begin{itemize}
-  \item $\tty$ is the tuple type with member types $\tys$;
-  \item \texttt{tys'} is the result of annotating all the types in
-    $\tys$;
-  \item \texttt{new\_ty} is the tuple type with member types \texttt{tys'}.
+  \item $\tty$ is the tuple type with member types $\tys$, that is, $\TTuple(\tys)$;
+  \item $\tys$ is the list $\tty_i$, for $i=1..k$;
+  \item annotating each type $\tty_i$ in $\tenv$, for $i=1..k$,
+  yields $\ttyp_i$ \ProseOrTypeError;
+  \item $\newty$ is the tuple type with member types $\ttyp$, for $i=1..k$.
 \end{itemize}
 
 \subsection{Example}
 In the following example, all the uses of tuple types are valid:
 \VerbatimInput{../tests/ASLTypingReference.t/TypingRule.TTuple.asl}
 
-
-
 \CodeSubsection{\TTupleBegin}{\TTupleEnd}{../Typing.ml}
 
+\begin{emptyformal}
+\subsection{Formally}
+\begin{mathpar}
+\inferrule{
+  \tys \eqname \tty_{1..k}\\
+  i=1..k: \annotatetype{\False, \tenv, \tty_i} \typearrow \ttyp_i \OrTypeError
+}{
+  \annotatetype{\Ignore, \tenv, \TTuple(\tys)} \typearrow \TTuple(\tysp)
+}
+\end{mathpar}
+\end{emptyformal}
 
 \isempty{\subsection{Comments}}
 
-
 \section{TypingRule.TArray \label{sec:TypingRule.TArray}}
 
 \subsection{Prose}
-Annotating a type~$\tty$ in an environment~\texttt{env} results in a
-rewritten type~\texttt{new\_ty} and all of the following apply:
+All of the following apply:
 \begin{itemize}
-  \item All of the following holds:
+  \item $\tty$ is the array type indexed by integer bounded by the
+    expression $\ve$ and of elements of type $\vt$;
+  \item Annotating the type $\vt$ in $\tenv$ yields $\vtp$ \ProseOrTypeError;
+  \item One of the following applies:
+  \begin{itemize}
+    \item All of the following apply (\textsc{expr\_is\_enum}):
+    \begin{itemize}
+      \item determining whether $\ve$ corresponds to an enumeration in $\tenv$
```

Comment:
the style here and below is markedly different from what is above - can this be minimised to help readers please?

Reply:
How is it different?